### PR TITLE
Add explanation of main to rustdoc docs

### LINKED ISSUE #17554: rustdoc: Document how and when rustdoc wraps a code block in a main function.
Currently, there is no indication in the documentation for rustdoc that it wraps code blocks in `fn main() { ... }`, nor how to stop it from doing so.

To give an example of how this can be very confusing, if you try to test the following code block, you get an error stating "unresolved import `f1`.  There is no `f1` in `???`" which is (seemingly) not true.

```
pub fn f1() {}

mod a {
	use f1;
	fn f2() { f1() }
}
```

Preferably, the docs should note that, by default, all code blocks are wrapped in a main function, that this can be avoided by ensuring the string "fn main" appears (or whatever the actual test is; I *think* it just looks for a string in the input), and that `#` can be used to hide this from the rendered output (which is already in the docs in a different context).

### PR DESCRIPTION
**Pull request by steveklabnik**
Fixes #17554

### CODE REVIEW

**Comment by alexcrichton** [comment]
The [full logic](https://github.com/rust-lang/rust/blob/055cc2ee742e29e389c5bd1da3a9d49ff8ca3724/src/librustdoc/test.rs#L217-L247) is actually a bit more complicated than this, but I'm not sure how much in detail it should be spelled out. Specifically, the steps taken are:

1. Given a code block, if it does not contain `fn main`, it is wrapped in `fn main() { <codeblock> }`
2. Given that result, if it contains no `extern crate` directives but it also contains the name of the crate being tested, then `extern crate <name>` is injected at the top.
3. Some common `allow` attributes are added for doc examples at the top.

This result is then tested.

Specifically, the example here isn't quite right, as what's actually tested look more like:

```
fn main() {
    use std::rc:Rc;
    let five = Rc::new(5);
}
```

**Comment by steveklabnik** [comment]
@alexcrichton I added the full thing, with some copyedit and formatting changes. I'm fine with just repeating it verbatim, I don't think this is incredibly likely to change.

File: src/doc/rustdoc.md [hunk 1]
```diff
@@ -217,6 +217,35 @@ spawn(move || { fib(200); })
...
+/// let five = Rc::new(5);
+/// ```
+```
+
+This will end up testing:
+
+```
+
+fn main() {
+    use std::rc:Rc;
```
Comment (eddyb) [inline src/doc/rustdoc.md]: `rc:Rc` should be `rc::Rc`.

File: src/doc/rustdoc.md [hunk 2]
```diff
@@ -217,6 +217,35 @@ spawn(move || { fib(200); })
...
 
+Rustdoc will automatically add a `main()` wrapper around your code, and in the right
+place. For example:
+
+```
+/// ```
+/// use std::rc::Rc;
+///
+/// let five = Rc::new(5);
+/// ```
```
Comment (eddyb) [inline src/doc/rustdoc.md]: Missing item after doc. @huonw suggests `# fn foo() {}`.

Comment (steveklabnik) [inline src/doc/rustdoc.md]: oh that's weird. This actually does work when processing doc comments though...

Comment (steveklabnik) [inline src/doc/rustdoc.md]: oh i'm an idiot

**Pull request merged**